code/layout-analyzer.js

```diff
@@ -31,32 +31,101 @@ window.addEventListener('DOMContentLoaded', () => {
 
   const charToKeys = char => keyChars[char] ?? keyChars[substituteChars[char]];
 
+  const is1DFH = keyCode =>
+    keyCode.startsWith('Key') ||
+      ['Space', 'Comma', 'Period', 'Slash', 'Semicolon'].includes(keyCode);
+
   // create an efficient hash table to parse a text
   const supportedChars = (keymap, deadkeys) => {
     const charTable = {};
     const deadTable = {};
 
-    // main chars, directly accessible (or with Shift / AltGr)
-    Object.entries(keymap).forEach(([key, values]) => {
-      values.forEach(char => {
-        if (!(char in charTable)) {
-          if (char.length === 1) {
-            charTable[char] = [key];
-          } else if (!(char in deadTable)) {
-            deadTable[char] = key;
-          }
-        }
-      });
-    });
+    // In case there are multiple ways of typing a singel char, this checks
+    // which sequence is easier to type (examples are in Ergo‑L)
+    const requiresLessEffort = (originalKeySequence, newKeySequence) => {
+      const uses1DK = '**' in keyboard.layout.deadKeys
+          ? keySequence => keySequence.some(key => key == charTable['**'][0])
+          : (_) => false;
+
+      const arrayCount = (array, predicate) => {
+        let rv = 0;
+        array.forEach(elem => { if (predicate(elem)) rv++ });
+        return rv;
+      };
```

Comment:
Out of curiosity, isn’t it possible to use `.reduce` here instead ?

Reply:
I mean, you *could* but it would be pretty counter intuitive, I feel. `reduce` is typically used when merging an array into a single value, and if we need extra transformations before that merge, that’s the role of `map`, `flatmap`, `filter`… (I used reduce for complex transformations in `computeNgrams` and I feel dirty)

Speaking of `filter`, I’m just realizing that a more FP way of implementing this would just be this : 

```js
const arrayCount = (array, predicate) => array.filter(predicate).length
```

Simple, concise, explicit. Everything I love about FP (though since `filter` creates a new array, it’s technically slower)

Reply:
I think it’s the one thing that you miss when writing FP, it’s that it’s not about having a single function that does everything in one go (like you often do with a 15 lines lambda in a `array.forEach` that mutates variables outside of it’s scope ; like dude, imperative is a perfectly fine style if you need a `for-loop` just use one 😛), but it’s more about composing smaller functions together to form pipelines that you feed your data through.

Reply:
I totally trust your judgement on that. And yes, the code is very readable as it is.